Assignment-5

smxlikith opened the pull request:
This pull request introduces a complete implementation of a Dynamic Inventory Management System in C, using dynamic memory allocation to efficiently manage product data. The system provides a menu-driven interface for core inventory operations, robust input validation, and modular code organization for maintainability and scalability.

**Core Inventory Management Functionality**
* Added `dynamicInventoryManagementSystem.c` with a menu-driven interface for adding, viewing, updating, searching, and deleting products, leveraging dynamic memory allocation for flexible inventory size.
* Implemented `lib/product.c` and `lib/product.h` to encapsulate product operations (add, view, update, search, delete) and memory management, including safe resizing and freeing of inventory.

**Robust Input Handling and Validation**
* Created `lib/userInput.c` and `lib/userInput.h` for safe and validated input of integers, floats, and strings, including range checks and buffer management to prevent invalid or unsafe input. 

**Configuration and Constraints**
* Added `lib/constraints.h` to define all configurable limits for product attributes and input validation, enabling easy adjustment of system parameters.

**Type Safety and Status Reporting**
* Introduced `lib/status.h` for standardized success/failure status codes used throughout the codebase for error handling and flow control.

**Documentation**
* Provided a comprehensive `README.md` with project overview, features, usage instructions, and compilation steps.

smxlikith committed:
Completed Assignment-5

tejas0123 commented on Assignment-5/lib/constraints.h:
```diff
@@ -0,0 +1,21 @@
+#ifndef CONTRAINTS_H
```
This can be clubbed with the userInput.h file

smxlikith replied:
I have separated the constraints to make them easier to modify without having to search for them in the code. I’ll merge them into userInput.h in the next commit

tejas0123 commented on Assignment-5/lib/status.h:
```diff
@@ -0,0 +1,9 @@
+#ifndef STATUS_H
```
Club with the userInput.h file.

smxlikith replied:
Yes, this can be merged into userInput.h. I’ve separated this into status.h for convenience

tejas0123 commented on Assignment-5/lib/product.c:
```diff
@@ -0,0 +1,216 @@
+#include <stdlib.h>
+#include <string.h>
+
+void init(int* size, Product** inventory) {
```
void is not an appropriate return type for this. Return the pointer to the memory block created

smxlikith replied:
since I am using a pointer to pointer for inventory, i am passing its reference here, so any changes made to inventory will be reflected in main. for resizing or memory allocation of the inventory array, it will be handled by the handleMemoryAllocation function

tejas0123 commented on Assignment-5/dynamicInventoryManagementSystem.c:
```diff
@@ -0,0 +1,56 @@
+    int currentInventorySize = 0;
+    Product* inventory = NULL;
+    init(&currentInventorySize, &inventory);
+    viewAllProducts(&currentInventorySize, &inventory);
```
Why pass by reference?

smxlikith replied:
since all the product related functions are designed to be run by the inventoryManagementMenu function, I declared them to receive references and handle dereferencing within the functions. Now, since i am using this  function outside the menu context, i am passing the reference directly.

tejas0123 commented on Assignment-5/dynamicInventoryManagementSystem.c:
```diff
@@ -0,0 +1,56 @@
+    init(&currentInventorySize, &inventory);
+    viewAllProducts(&currentInventorySize, &inventory);
+    while (1) {
+        runInventoryMenu(&currentInventorySize, &inventory);
```
Same here

smxlikith replied:
runInventoryMenu requires a reference to both the size and the inventory because it calls functions that may modify the inventory and its size

tejas0123 commented on Assignment-5/dynamicInventoryManagementSystem.c:
```diff
@@ -0,0 +1,56 @@
+                searchProductByName(currentSize, inventory);
+                break;
+            case 6:
+                searchProductByPriceRange(currentSize, inventory);
```
When do you need to actually pass by reference? Are you modifying the actual memory block's pointer itself here?

smxlikith replied:
since i am passing the reference to the array in the runInventoryMenu function parameters, I’ve kept it as is and handle the dereferencing within the functions themselves.

tejas0123 replied:
Pass by reference only to the functions where you are actually modifying the memory block using realloc. Update, search and other functions do not need double pointer arguments

tejas0123 commented on Assignment-5/lib/product.c:
```diff
@@ -0,0 +1,216 @@
+    int ID;
+    printf("Enter new product details:\n");
+    if (handleMemoryAllocation(*currentSize + 1, inventory) == SUCCESS) {
+        while (1) {
```
use while(true)

smxlikith replied:
since c doesn't support boolean by default, I have used int

tejas0123 replied:
stdbool.h

tejas0123 commented on Assignment-5/lib/userInput.c:
```diff
@@ -0,0 +1,153 @@
+    return SUCCESS;
+}
+
+void getProductId(int* IN) {
+    while (1) {
+        printf("Product ID: ");
+        if (getInt(IN) == SUCCESS) {
+            if (*IN >= MIN_PRODUCT_ID && *IN <= MAX_PRODUCT_ID) {
+                break;
+            }
+            printf("Input out of range. ID must be %d <= ID <= %d, but got %d.\n", MIN_PRODUCT_ID, MAX_PRODUCT_ID, *IN);
+        }
+        printf("Invalid input, Try again...\n");
+    }
+}
+
+void getProductName(char* IN) {
+    while (1) {
+        printf("Product Name: ");
+        if (getString(IN) == SUCCESS) {
+            break;
+        }
+    }
+}
+
+void getProductPrice(float* IN) {
+    while (1) {
+        printf("Product price: ");
+        if (getFloat(IN) == SUCCESS) {
```
These functions violate SRP. The name indicates user input, but you are also assigning it to memory. Combine this logic with the calling functions. Also create functions for logic which is reused in multiple places. Like id, name

tejas0123 commented on Assignment-5/lib/product.c:
```diff
@@ -0,0 +1,216 @@
+    return SUCCESS;
+}
+
+Status searchProductId(int ID, int* currentSize, Product** inventory) {
```
This function should return index. Success or failure should be returned by functions which are doing some operation or computation that denotes ends up as success or failure. If you are adding / updating / deleting a product you should return success or failure. But if you are transforming some data or manipulating data or searching for some data / geting data as input from user, you should return the actual result / data you get. Consider these points and make changes in all the functions wherever necessary.

smxlikith replied:
since i am searching for the product ID, if the search is successful, i know the product already exists in the array. when adding details for a new product, i use the success/failure values to determine if it exists in the array. failure occurs when the product ID isn’t found or if the inventory is empty.

tejas0123 commented on Assignment-5/lib/product.c:
```diff
@@ -0,0 +1,216 @@
+    return FAILURE;
+};
+
+Status viewAllProducts(int* currentSize, Product** inventory) {
```
This function's purpose is to print all the products. After printing are we bothered whether it was successful or not? Status should be returned only for computations or tasks done

tejas0123 replied:
Also you can see here that you only return success. There is no path where you are returning failure. That shows that the return type is not appropriate.

tejas0123 replied:
Printing, logging functions, in other words any which does not produce actual results that can be returned can be made void. Usually the last function in the chain of function calls will be void

smxlikith replied:
yes, you are right, this could be set to void

tejas0123 commented:
Input handling and validation has been implemented very well. 👍🏼

tejas0123 commented:
- Pass by reference only to the functions where you are actually modifying the memory block using realloc. Update, search and other functions do not need double pointer arguments
- Use stdbool.h in the future for all boolean variables

smxlikith committed:
updated the code:
	- merged status and constraints headers into the userInput header.
	- using stdbool instead of directly using int values.
	- updated functions
